### PR TITLE
Api closing

### LINKED ISSUE #5791: [Next Major] Close API
- [ ] Add `final` modifier
- [ ] Reduce visiblity to `private`
- [ ] Add type hints

### PR DESCRIPTION
**Pull request by core23**
## Subject

I am targeting this branch, because this is a BC break.

Closes #5791

## Changelog

```markdown
### Changed
- Add final modifier to classes 
- Reduce visibility of methods and properties
```

## To do
    
- [ ] Check final classes and introduce interfaces / make all API methods final

### CODE REVIEW

File: tests/Action/AppendFormFieldElementActionTest.php [hunk 1]
```diff
@@ -62,7 +62,7 @@ final class AppendFormFieldElementActionTest extends TestCase
      */
     private $helper;
 
-    protected function setUp(): void
+    private function setUp(): void
     {
```
Comment (OskarStark) [inline tests/Action/AppendFormFieldElementActionTest.php]: Is this correct?

Comment (core23) [reply tests/Action/AppendFormFieldElementActionTest.php]: Fixed

**Comment by core23** [comment]
How should we handle classes without an interface? Should we mark them as final. If yes, we can't mock these classes in PHPUnit

**Comment by greg0ire** [comment]
> How should we handle classes without an interface? Should we mark them as final. If yes, we can't mock these classes in PHPUnit

Maybe let's mark them as final if they are easy not to mock (no dependencies to classes that do IO, and few dependencies)?

**Comment by VincentLanglet** [comment]
I'm curious, how was decided the class which should be final and the class which should not ?

For example, I extends the `Sonata\DoctrineORMAdminBundle\Model\ModelManager` or the `Sonata\Exporter\Source\DoctrineORMQuerySourceIterator` in my project. I know it's not from the AdminBundle but should I be worried that these class will be one day final ?

**Comment by core23** [comment]
> I'm curious, how was decided the class which should be final and the class which should not ?

We will be marking everything as final by default unless someone will come up with a very good reason, why there should be an extension point. This will help us in maintaining this project and allows us to do some things that otherwise will result in a BC break.

**Comment by core23** [comment]
I reverted all conflicting files for now

File: src/Resources/skeleton/Admin.tpl.php [hunk 2]
```diff
@@ -13,13 +13,13 @@
 final class <?= $class_name ?> extends AbstractAdmin
```
Comment (VincentLanglet) [inline src/Resources/skeleton/Admin.tpl.php]: Why the admin are final by default ?
In our project we often extends our admin, and I feel like that making the class final by default is like saying "Extending an admin is a bad practice".

File: src/Resources/skeleton/Admin.tpl.php [hunk 3]
```diff
@@ -13,13 +13,13 @@
...
 {
 
-    protected function configureDatagridFilters(DatagridMapper $datagridMapper): void
+    private function configureDatagridFilters(DatagridMapper $datagridMapper): void
```
Comment (VincentLanglet) [inline src/Resources/skeleton/Admin.tpl.php]: The function is protected in the AbstractAdmin, shouldn't this stay protected then ?

**Pull request closed without merging**